// vite-preset-react/dist/index.js
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};

// src/index.ts
import react from "@vitejs/plugin-react";

// src/devtools.ts
function reactDevtoolsPlugin({
  removeInProd = false
} = {}) {
  const plugin = {
    name: "react:devtools",
    enforce: "pre",
    apply: "build",
    transformIndexHtml(code) {
      if (removeInProd) {
        return {
          html: code,
          tags: [
            {
              injectTo: "body",
              tag: `script`,
              children: `if (typeof window.__REACT_DEVTOOLS_GLOBAL_HOOK__ === 'object') { window.__REACT_DEVTOOLS_GLOBAL_HOOK__.inject = function () {};};`
            }
          ]
        };
      }
      return code;
    }
  };
  return plugin;
}

// src/index.ts
function reactPlugin({
  removeDevtoolsInProd = false,
  injectReact = true,
  reactPluginOptions: reactRefreshOptions
} = {}) {
  return [
    react(__spreadValues({ jsxRuntime: injectReact ? "automatic" : "classic" }, reactRefreshOptions)),
    reactDevtoolsPlugin({ removeInProd: removeDevtoolsInProd })
  ];
}
export {
  reactPlugin as default
};
